nonlinear_solid_mechanics/nonlinear_solid_mechanics.cc

```diff
@@ -0,0 +1,1495 @@
+
+    AffineConstraints<double> constraints;
+    BlockSparsityPattern      sparsity_pattern;
+    BlockSparseMatrix<double> tangent_matrix;
+    BlockVector<double>       system_rhs;
+    BlockVector<double>       total_displacement;
+    BlockVector<double>       total_displacement_old;
+    BlockVector<double>       velocity;
+    BlockVector<double>       velocity_old;
+    BlockVector<double>       acceleration;
+    BlockVector<double>       acceleration_old;
+
+    // Alias
+    std::vector<BlockVector<double> *> state_variables;
+
+    // container for data exchange with precice
+    BlockVector<double> external_stress;
+
+    PreciceDealCoupling::CouplingFunctions<dim, BlockVector<double>>
+      coupling_functions;
+
+    struct Errors
+    {
+      Errors()
+        : u(1.0)
+      {}
+
+      void
+      reset()
+      {
+        u = 1.0;
+      }
+      void
+      normalise(const Errors &val)
+      {
+        if (val.u != 0.0)
+          u /= val.u;
+      }
+
+      double u;
+    };
+
+    Errors error_residual, error_residual_0, error_residual_norm, error_update,
+      error_update_0, error_update_norm;
+
+    void
+    get_error_residual(Errors &error_residual);
+
+    void
+    get_error_update(const BlockVector<double> &newton_update,
+                     Errors &                   error_update);
+
+    static void
+    print_conv_header();
+
+    void
+    print_conv_footer();
+  };
+
+
+
+  template <int dim, typename NumberType>
+  Solid<dim, NumberType>::Solid(const std::string &case_path)
+    : parameters(Parameters::AllParameters(case_path + "parameters.prm"))
+    , vol_reference(0.0)
+    , vol_current(0.0)
+    , triangulation(Triangulation<dim>::maximum_smoothing)
+    , time(parameters.end_time, parameters.delta_t)
+    , timer(std::cout, TimerOutput::summary, TimerOutput::wall_times)
+    , degree(parameters.poly_degree)
+    , fe(FE_Q<dim>(parameters.poly_degree), dim)
+    , // displacement
+    dof_handler_ref(triangulation)
+    , dofs_per_cell(fe.dofs_per_cell)
+    , u_fe(first_u_component)
+    , dofs_per_block(n_blocks)
+    , qf_cell(parameters.poly_degree + 2)
+    , qf_face(parameters.poly_degree + 2)
+    , n_q_points(qf_cell.size())
+    , n_q_points_f(qf_face.size())
+    , coupling_functions(parameters)
+  {}
+
+  template <int dim, typename NumberType>
+  Solid<dim, NumberType>::~Solid()
+  {
+    dof_handler_ref.clear();
+  }
+
+
+  template <int dim, typename NumberType>
+  void
+  Solid<dim, NumberType>::run()
+  {
+    make_grid();
+    system_setup();
+    output_results();
+    time.increment();
+
+    coupling_functions.initialize_precice(dof_handler_ref,
+                                          total_displacement,
+                                          external_stress);
+
+
+    BlockVector<NumberType> solution_delta(dofs_per_block);
+    while ((time.current() <= time.end()) &&
+           coupling_functions.precice.isCouplingOngoing())
```

Comment:
```suggestion
    while (coupling_functions.precice.isCouplingOngoing())
```
I think it's educationally better like this. Readers need to understand that the control over the end of the simulation is given to preCICE.